src/aes.rs

```diff
@@ -16,316 +24,157 @@ impl Default for Aes {
-            );
-            aes.wo(
-                aes::KEYWR4,
-                u32::from_ne_bytes(key[16..20].try_into().unwrap()),
-            );
-            aes.wo(
-                aes::KEYWR5,
-                u32::from_ne_bytes(key[20..24].try_into().unwrap()),
-            );
-            aes.wo(
-                aes::KEYWR6,
-                u32::from_ne_bytes(key[24..28].try_into().unwrap()),
-            );
-            aes.wo(
-                aes::KEYWR7,
-                u32::from_ne_bytes(key[28..32].try_into().unwrap()),
-            );
-
-            // Set the IV.
-            aes.wo(aes::IVR0, u32::from_ne_bytes(iv[0..4].try_into().unwrap()));
-            aes.wo(aes::IVR1, u32::from_ne_bytes(iv[4..8].try_into().unwrap()));
-            aes.wo(aes::IVR2, u32::from_ne_bytes(iv[8..12].try_into().unwrap()));
-            aes.wo(
-                aes::IVR3,
-                u32::from_ne_bytes(iv[12..16].try_into().unwrap()),
-            );
-
-            // Write the input data.
-            aes.wo(
-                aes::IDATAR0,
-                u32::from_ne_bytes(chunk[0..4].try_into().unwrap()),
-            );
-            aes.wo(
-                aes::IDATAR1,
-                u32::from_ne_bytes(chunk[4..8].try_into().unwrap()),
-            );
-            aes.wo(
-                aes::IDATAR2,
-                u32::from_ne_bytes(chunk[8..12].try_into().unwrap()),
-            );
-            aes.wo(
-                aes::IDATAR3,
-                u32::from_ne_bytes(chunk[12..16].try_into().unwrap()),
-            );
-
-            // Start the encryption process.
-            aes.wfo(aes::CR_START, 1);
-
-            // Wait until the encryption is done.
-            while aes.rf(aes::ISR_DATRDY) == 0 {}
+    /// Initialize the AES peripheral in CBC mode.
+    pub fn init(&self, key: [u32; 4], iv: [u32; 4]) {
+        self.reset();
+        let mut csr = CSR::new(self.base_addr as *mut u32);
+        self.set_key(&key);
+        self.set_iv(&iv);
+        let opmod = csr.ms(MR_OPMOD, 1); // CBC mode
+        let smod = csr.ms(MR_SMOD, 2); // DMA auto-start
+        let ckey = csr.ms(MR_CKEY, CKEY);
+        // TODO Try disabling this?
+        let dualbuff = csr.ms(MR_DUALBUFF, 1);
+        csr.wo(MR, ckey | opmod | smod | dualbuff);
+    }
 
-            // Copy the output data.
-            let start = i * 16;
-            chunk[start + 0..start + 4].copy_from_slice(&aes.r(aes::ODATAR0).to_ne_bytes());
-            chunk[start + 4..start + 8].copy_from_slice(&aes.r(aes::ODATAR1).to_ne_bytes());
-            chunk[start + 8..start + 12].copy_from_slice(&aes.r(aes::ODATAR2).to_ne_bytes());
-            chunk[start + 12..start + 16].copy_from_slice(&aes.r(aes::ODATAR3).to_ne_bytes());
+    /// Encrypt some data using the AES peripheral. Panics if the output buffer is smaller
+    /// than the input buffer.
+    pub fn encrypt(
+        &self,
+        input_phys: &[u8],
```

Comment:
Note: I dislike using slices for physical addresses, because it's too easy to just give a slice in virtual address space to the function.

Using slices for physical (invalid) addresses is kind of unpredictable too because rust uses some pretty aggressive optimizations around them with the strict non-aliasing rules. It shouldn't be a problem here, it's just generally unclean.

Maybe we should introduce a physical address and physical slice type sometime and use that universally.

Reply:
Added something in https://github.com/Foundation-Devices/atsama5d27/pull/46/commits/e109f84b014c20eeb6aa9a0a5939a598a4d97893, lmk what do you think

Reply:
Perfect.